### PR TITLE
Modify the DataGridView control to update the strong reference to wea…

### LINKED ISSUE #9224: Update AccessibleObject owner references weak references that were missed in PR#9164
Update AccessibleObject owner references weak references that were missed in pr#[9164](https://github.com/dotnet/winforms/pull/9164)
After the AccessibleObject  was updated, the corresponding test case "ControlAccessibleObject_DoesNotRootControls_AllPublicControl" added from pr#[9215 ](https://github.com/dotnet/winforms/pull/9215) also need to be updated. 

- [x] CheckedListBox #9324 
- [x] ComboBox #9323 
- [x] DataGridView #9376 
- [x] DataGridViewComboBoxEditingControl #9376 
- [x] DataGridViewTextBoxEditingControl #9376 
- [x] FlowLayoutPanel PR #9246 
- [x] HScrollBar PR #9434 
- [x] Label  PR #9246 
- [x] LinkLabel PR #9246 
- [x] ListBox #9324 
- [x] ListView #9451 
- [x] MaskedTextBox PR #9400 
- [x] MonthCalendar #9454 
- [x] Panel PR #9246 
- [x] PropertyGrid #9452 
- [x] TableLayoutPanel PR #9246 
- [x] TabPage #9453 
- [x] TextBox #9400 
- [x] ToolStripContentPanel PR #9246 
- [x] TreeView #9458 
- [x] VScrollBar PR #9434

### PR DESCRIPTION
**Pull request by SimonZhao888**
Related to #9224

## Proposed changes

- Update the control owning field of DataGridView to using weak references
- Enable the GC collect test for the DataGridView  control in unit test cases

## Customer Impact

- The strong reference used by the original Owner property can cause a memory leak

## Regression? 

- No

## Test methodology 

- Unit tests

## Test environment(s) 

- 8.0.100-preview.7.23324.2

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9376)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs [hunk 1]
```diff
@@ -56,85 +56,79 @@ public override AccessibleRole Role
...
-                AccessibleRole role = _ownerDataGridView.AccessibleRole;
-                if (role != AccessibleRole.Default)
-                {
-                    return role;
-                }
-
                 // the Default AccessibleRole is Table
-                return AccessibleRole.Table;
+                return _ownerDataGridView.TryGetTarget(out DataGridView? owner) ? owner.AccessibleRole : AccessibleRole.Table;
             }
```
Comment (v-elnovikova) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:61]: If owner's `AccessibleRole` is not set explicitly, this would return `AccessibleRole.Default`, not `AccessibleRole.Table`. Is it intended?

Comment (v-elnovikova) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:61]: [GetOwnerAccessibleRole](https://github.com/dotnet/winforms/blob/dbf5345ea92926c9f0d9ba315ce4d45b678292e1/src/System.Windows.Forms/src/System/Windows/Forms/AccessibleObjectExtensions.cs#L307) extension method might be used here:

```
     public override AccessibleRole Role
            => this.GetOwnerAccessibleRole(AccessibleRole.Table);
```

**Review by v-elnovikova — APPROVED** [review]
Looks good!

File: src/System.Windows.Forms/src/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -1 +1,2 @@
 override System.Windows.Forms.ToolStripButton.ProcessDialogKey(System.Windows.Forms.Keys keyData) -> bool
+override System.Windows.Forms.DataGridView.DataGridViewAccessibleObject.GetSelected() -> System.Windows.Forms.AccessibleObject?
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/PublicAPI.Unshipped.txt:2]: @dreddy-work - can we make changes like that?

**Review by dkazennov — APPROVED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs [hunk 3]
```diff
@@ -14,20 +14,20 @@ protected class DataGridViewAccessibleObject : ControlAccessibleObject
         private int[]? runtimeId; // Used by UIAutomation
         private bool? _isModal;
 
-        private readonly DataGridView _ownerDataGridView;
+        private readonly WeakReference<DataGridView> _ownerDataGridView;
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:17]: @SimonZhao888  - is there anything that's preventing from removing this field and using `TryGetOwnerAs` in this PR?

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs [hunk 4]
```diff
@@ -53,88 +53,76 @@ internal void ReleaseChildUiaProviders()
...
             }
         }
 
-        private AccessibleObject SelectedCellsAccessibilityObject
+        private AccessibleObject? SelectedCellsAccessibilityObject
         {
             get
             {
-                _selectedCellsAccessibilityObject ??= new DataGridViewSelectedCellsAccessibleObject(_ownerDataGridView);
+                _selectedCellsAccessibilityObject ??= _ownerDataGridView.TryGetTarget(out DataGridView? owner) ? new DataGridViewSelectedCellsAccessibleObject(owner) : null;
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:72]: Can we pass `DataGridViewAccessibleObject` as an argument to `DataGridViewSelectedCellsAccessibleObject` constructor, to avoid returning null in `SelectedCellsAccessibilityObject` and thus changing public API signature? In `DataGridViewSelectedCellsAccessibleObject` we can then do something like `_owningDataGridViewSelectedCellsAccessibleObject.TryGetOwnerAs(out DataGridView? owner)` to get owning DataGridView.

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs]: Thanks for your reminder, I had updated it, and please help me review it again.

File: src/System.Windows.Forms/src/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/PublicAPI.Unshipped.txt]: Hi @Tanya-Solyanik , under the guidance of dmitrii, I have modified the DataGridViewSelectedCellsAccessibleObject class and have now restored the part of the public API that was changed.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs [hunk 5]
```diff
@@ -257,26 +246,31 @@ internal override int[] RuntimeId
 
         internal override object? GetPropertyValue(UiaCore.UIA propertyID)
         {
+            if (!this.TryGetOwnerAs(out DataGridView? owner))
+            {
+                return null;
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:251]: I would suggest to get owner inside every case because some properties don't require it. For example, if owner is null, IsControlElementPropertyId will now return null instead true. Same with properties in base class.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs [hunk 6]
```diff
@@ -8,11 +8,11 @@ public partial class DataGridView
 {
     private class DataGridViewSelectedCellsAccessibleObject : AccessibleObject
     {
-        private readonly DataGridView _ownerDataGridView;
+        private readonly DataGridViewAccessibleObject _ownerDataGridViewAO;
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs:11]: We usually spell out abbreviations
```suggestion
        private readonly DataGridViewAccessibleObject _ownerDataGridViewAccessibleObject;
```

File: src/System.Windows.Forms/src/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/PublicAPI.Unshipped.txt]: Great collaboration, guys!

**Comment by Tanya-Solyanik** [comment]
@SimonZhao888  - do you have permissions to see CI build errors?

```

Error message
Microsoft.CSharp.RuntimeBinder.RuntimeBinderException : 'System.Dynamic.DynamicObject' does not contain a definition for '_ownerDataGridView'



Stack trace
   at CallSite.Target(Closure, CallSite, Object)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
   at System.Windows.Forms.Tests.DataGridViewSelectedCellsAccessibleObjectTests.DataGridViewSelectedCellsAccessibleObject_Ctor_Default() in /_/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DataGridViewSelectedCellsAccessibleObjectTests.cs:line 19
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

File: src/System.Windows.Forms/src/PublicAPI.Unshipped.txt [hunk 7]
```diff
@@ -1 +1 @@
-override System.Windows.Forms.ToolStripButton.ProcessDialogKey(System.Windows.Forms.Keys keyData) -> bool
+override System.Windows.Forms.ToolStripButton.ProcessDialogKey(System.Windows.Forms.Keys keyData) -> bool
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/PublicAPI.Unshipped.txt:1]: Please undo this change.

Comment (SimonZhao888) [reply src/System.Windows.Forms/src/PublicAPI.Unshipped.txt:1]: Yes, I can see the build errors, and I'm working on it.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs (hunk 5 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:251]: @dmitrii-drobotov , could you help me look at the CI build errors? After the last commit, the test case has these errors. To fix these errors, I tried to add a new constructor to the DataGridViewSelectedCellsAccessibleObject class, as: 


        public DataGridViewSelectedCellsAccessibleObject(DataGridView owner)
        {
            _ownerDataGridViewAccessibleObject = (DataGridViewAccessibleObject)(owner.AccessibilityObject ?? owner.CreateAccessibilityInstance());
        }


But, there is a new error with the following error message: Ambiguous match found for 'System. System.Windows.Forms.DataGridView)'.
I have no idea how to solve this problem, could you help me?

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs (hunk 6 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs]: Done.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs (hunk 5 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs]: Done.

File: src/System.Windows.Forms/src/PublicAPI.Unshipped.txt (hunk 7 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/PublicAPI.Unshipped.txt:1]: In fact, I have removed the changes I made, and this content is added by PR https://github.com/dotnet/winforms/pull/9328

**Review by dkazennov — APPROVED** [review]

File: src/System.Windows.Forms/src/PublicAPI.Unshipped.txt (hunk 7 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/PublicAPI.Unshipped.txt:1]: This is probably a change in formatting or whitespaces, but still please undo it to keep the commit clean.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs [hunk 8]
```diff
@@ -8,11 +8,11 @@ public partial class DataGridView
 {
     private class DataGridViewSelectedCellsAccessibleObject : AccessibleObject
     {
-        private readonly DataGridView _ownerDataGridView;
+        private readonly DataGridViewAccessibleObject _ownerDataGridViewAccessibleObject;
 
-        public DataGridViewSelectedCellsAccessibleObject(DataGridView owner)
+        public DataGridViewSelectedCellsAccessibleObject(DataGridViewAccessibleObject owner)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs:13]: I think this will be confusing, because everywhere else we are using `owner` to represent control that is associated with the accessible object. Consider `parentAccessibleObject`.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs [hunk 9]
```diff
@@ -8,11 +8,11 @@ public partial class DataGridView
 {
     private class DataGridViewSelectedCellsAccessibleObject : AccessibleObject
     {
-        private readonly DataGridView _ownerDataGridView;
+        private readonly DataGridViewAccessibleObject _ownerDataGridViewAccessibleObject;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs:11]: Consider `_parentAccessibleObject`

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks good!

File: src/System.Windows.Forms/src/PublicAPI.Unshipped.txt (hunk 7 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/PublicAPI.Unshipped.txt]: Done.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs (hunk 8 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs]: I will pay more attention to these.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs (hunk 9 above)
Comment (SimonZhao888) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.SelectedCellsAccessibleObject.cs]: Done.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs [hunk 10]
```diff
@@ -257,26 +246,33 @@ internal override int[] RuntimeId
...
-                    return (IsModal || RowCount == 0) && _ownerDataGridView.Focused;
+                    return this.TryGetOwnerAs(out owner) && (IsModal || RowCount == 0) && owner.Focused;
                 case UiaCore.UIA.IsControlElementPropertyId:
                     return true;
                 case UiaCore.UIA.IsKeyboardFocusablePropertyId:
-                    return _ownerDataGridView.CanFocus;
+                    return this.TryGetOwnerAs(out owner) && owner.CanFocus;
                 case UiaCore.UIA.ItemStatusPropertyId:
                     var canSort = false;
+                    if(!this.TryGetOwnerAs(out owner))
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridView.AccessibleObject.cs:267]:
```suggestion
                    if (!this.TryGetOwnerAs(out owner))
```

**Review by Tanya-Solyanik — DISMISSED** [review]

**Review by dmitrii-drobotov — APPROVED** [review]

**Comment by dreddy-work** [comment]
@SimonZhao888, please resolve conflict.

**Review by Tanya-Solyanik — APPROVED** [review]

**Review by Tanya-Solyanik — APPROVED** [review]
